debezium-connector-mysql/src/main/java/io/debezium/connector/mysql/MySqlDatabaseSchema.java

```diff
@@ -155,22 +162,46 @@ public boolean isGlobalSetVariableStatement(String ddl, String databaseName) {
 
     @Override
     public void applySchemaChange(SchemaChangeEvent schemaChange) {
-        applySchemaChange(schemaChange, SchemaChangeEventConsumer.NOOP);
+        switch (schemaChange.getType()) {
+            case CREATE:
+            case ALTER:
+                schemaChange.getTableChanges().forEach(x -> buildAndRegisterSchema(x.getTable()));
+                break;
+            case DROP:
+                schemaChange.getTableChanges().forEach(x -> removeSchema(x.getId()));
+                break;
+            default:
+        }
+
+        // Record the DDL statement so that we can later recover them if needed. We do this _after_ writing the
+        // schema change records so that failure recovery (which is based on of the history) won't lose
+        // schema change records.
+        // We are storing either
+        // - all DDLs if configured
+        // - or global SET variables
+        // - or DDLs for monitored objects
+        if (!databaseHistory.storeOnlyMonitoredTables() || isGlobalSetVariableStatement(schemaChange.getDdl(), schemaChange.getDatabase())
+                || schemaChange.getTables().stream().map(Table::id).anyMatch(filters.dataCollectionFilter()::isIncluded)) {
+            LOGGER.debug("Recorded DDL statements for database '{}': {}", schemaChange.getDatabase(), schemaChange.getDdl());
+            record(schemaChange, schemaChange.getTableChanges());
+        }
     }
 
-    public void applySchemaChange(SchemaChangeEvent schemaChange, SchemaChangeEventConsumer schemaEventConsumer) {
-        LOGGER.debug("Applying schema change event {}", schemaChange);
+    public List<SchemaChangeEvent> parseSnapshotDdl(String ddlStatements, String databaseName, MySqlOffsetContext offset, Instant sourceTime) {
+        LOGGER.debug("Processing snapshot DDL '{}' for database '{}'", ddlStatements, databaseName);
+        return parseDdl(ddlStatements, databaseName, offset, sourceTime, true);
+    }
 
-        if (schemaChange.getType() != SchemaChangeEventType.RAW) {
-            LOGGER.debug("Schema change already processed by the database schema");
-            return;
-        }
+    public List<SchemaChangeEvent> parseStreamingDdl(String ddlStatements, String databaseName, MySqlOffsetContext offset, Instant sourceTime) {
+        LOGGER.debug("Processing streaming DDL '{}' for database '{}'", ddlStatements, databaseName);
+        return parseDdl(ddlStatements, databaseName, offset, sourceTime, false);
+    }
 
-        final String ddlStatements = schemaChange.getDdl();
-        final String databaseName = schemaChange.getDatabase();
+    private List<SchemaChangeEvent> parseDdl(String ddlStatements, String databaseName, MySqlOffsetContext offset, Instant sourceTime, boolean snapshot) {
+        final List<SchemaChangeEvent> schemaChangeEvents = new ArrayList<>();
```

Comment:
Let's allocate it with just 3 elements, should be enough for most DDL events.

Reply:
Done